src/axes.jl

```diff
@@ -178,16 +179,21 @@ for f in [:first, :last]
     @eval @inline Base.$f(r::IdOffsetRange) = eltype(r)($f(r.parent) + r.offset)
 end
 
-@inline function Base.iterate(r::IdOffsetRange)
-    ret = iterate(r.parent)
-    ret === nothing && return nothing
-    return (eltype(r)(ret[1] + r.offset), ret[2])
-end
-@inline function Base.iterate(r::IdOffsetRange, i)
-    ret = iterate(r.parent, i)
+# Iteration for an IdOffsetRange
+@inline Base.iterate(r::IdOffsetRange, i...) = _iterate(r, i...)
+# In general we iterate over the parent term by term and add the offset.
+# This might have some performance degradation when coupled with bounds-checking
+# See https://github.com/JuliaArrays/OffsetArrays.jl/issues/214
+@inline function _iterate(r::IdOffsetRange, i...)
+    ret = iterate(r.parent, i...)
     ret === nothing && return nothing
     return (eltype(r)(ret[1] + r.offset), ret[2])
 end
+# Base.OneTo(n) is known to be exactly equivalent to the range 1:n,
+# and has no specialized iteration defined for it,
+# so we may add the offset to the range directly and iterate over the result
+# This gets around the performance issue described in issue #214
+@inline _iterate(r::IdOffsetRange{<:Integer, <:Base.OneTo}, i...) = iterate(r.parent .+ r.offset, i...)
```

Comment:
I'm not sure of it, `r.parent .+ r.offset` is a little bit tricky to me, maybe manually expand it?

```suggestion
@inline function _iterate(r::IdOffsetRange{T, <:Base.OneTo}, i...) where T
    r0 = UnitRange{T}(r.offset + 1, r.offset + r.stop)
    iterate(r0, i...)
end
```

Reply:
`OneTo` has special broadcasting methods defined to cover this, so I don't think this will make much of a difference. Are you concerned about the behavior changing?

Reply:
Yeah, this is expected to be identical, but the current improvement is perhaps a compiler trick so I'm wondering if a manual inline/expand helps stabilize this optimization.

Reply:
Makes sense. I'll add some util functions that might come in useful at other points as well.